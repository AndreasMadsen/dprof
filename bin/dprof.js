#!/bin/env node

var endpoint = require('endpoint');
var extend = require('util-extend');

process.stdin.pipe(endpoint(function (err, dprof) {
  if (err) throw err;
  dprof = JSON.parse(dprof);

  var content = new Node(dprof.root, dprof.total, 1, new State(), new Root());

  var svg = new Element('svg').attr({ xmlns: 'http://www.w3.org/2000/svg', width: 1000 });

  process.stdout.write(svg.start() + '\n');
  process.stdout.write(content.stringify() + '\n');
  process.stdout.write(svg.end());
}));

function Root() {
  this.top = 0;
}

function State() {
  this.y = 0;
}

function Node(node, total, depth, state, parent) {
  this.name = node.name;
  this.init = Math.floor(node.init / total * 1000);
  this.before = Math.floor((node.init + node.before) / total * 1000);
  this.after = Math.floor((node.init + node.after) / total * 1000);
  this.stack = node.stack;

  this.depth = depth;
  this.state = state;
  this.parent = parent;
  this.top = state.y;

  this.elems = [];
  this.elems.push(new Element('rect').attr({
    x: this.init,
    y: this.top,
    width: this.before - this.init,
    height: 20,
    fill: 'SteelBlue',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('rect').attr({
    x: this.before,
    y: this.top,
    width: this.after - this.before,
    height: 20,
    fill: 'IndianRed',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('line').attr({
    x1: this.init,
    y1: this.parent.top + 20,
    x2: this.init,
    y2: this.top + 20,
    style: 'stroke: #000; stroke-width: 2px'
  }));

  this.state.y += 20;
  this.children = node.children.map(function (child) {
    return new Node(child, total, depth + 1, state, this);
  }, this);
}

Node.prototype._indent = function () {
  var s = '';
  for (var i = 0; i < this.depth; i++) {
    s += '  ';
  }
  return s;
};

Node.prototype.stringify = function () {
  var content = [];
  this.elems.forEach(function (elem) {
    content.push(this._indent() + elem.singleton());
  }, this);
  this.children.forEach(function (child) {
    content.push(child.stringify());
  }, this);

  return content.join('\n');
};

function Element(name) {
  this._name = name;
  this._attr = {};
}

Element.prototype.attr = function (attr) {
  extend(this._attr, attr);
  return this;
};

Element.prototype.start = function () {
  var start = '<' + this._name;
  Object.keys(this._attr).forEach(function (k) {
    start += ' ' + k + '="' + this._attr[k] + '"';
  }, this);
  start += '>';
  return start;
};

Element.prototype.end = function () {
  return '</' + this._name + '>';
};

Element.prototype.singleton = function () {
  return this.start() + this.end();
};
