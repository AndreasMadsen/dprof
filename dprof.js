'use strict';

const providers = process.binding('async_wrap').Providers;
const asyncWrap = require('./async_wrap.js');
const zlib = require('zlib');
const fs = require('fs');

const version = require('./package.json').version;
const processStart = process.hrtime();

// Change the default stackTraceLimit if it haven't allready been overwritten
if (process.execArgv.indexOf('--stack_trace_limit') === -1 && Error.stackTraceLimit === 10) {
  Error.stackTraceLimit = 8;
}

//
// Define node class
//

function Site(site) {
  this.description = site.toString();
  this.filename = site.getFileName();
  this.line = site.getLineNumber();
  this.collum = site.getColumnNumber();
}

function timestamp() {
  const t = process.hrtime(processStart);
  return t[0] * 1e9 + t[1];
}

function Node(name, stack) {
  this.name = name;
  this._init = timestamp();
  this._before = [];
  this._after = [];
  this.children = [];
  this.stack = stack.map(function (site) {
    return new Site(site);
  });
}

Node.prototype.add = function (handle) {
  const node = new Node(handle.constructor.name, asyncWrap.stackTrace(3));
  this.children.push(node);
  return node;
};

Node.prototype.before = function () {
  this._before.push(timestamp());
};

Node.prototype.after = function () {
  this._after.push(timestamp());
};

Node.prototype.toJSON = function () {
  return {
    name: this.name,
    init: this._init,
    before: this._before,
    after: this._after,
    stack: this.stack,
    children: this.children
  };
};

Node.prototype.rootIntialize = function () {
  this._init = 0;
  this._before.push(0);
};

//
// Setup hooks
//

asyncWrap.setup(asyncInit, asyncBefore, asyncAfter);

const root = new Node('root', asyncWrap.stackTrace(2));
      root.rootIntialize();
let state = root;

function asyncInit(provider, parent) {
  if (provider === providers.TIMERWRAP) {
    this._dprofIgnore = true;
    return;
  }

  if (parent) {
    this._dprofState = parent._dprofState.add(this, provider, parent);
  } else {
    this._dprofState = state.add(this, provider, parent);
  }
}

function asyncBefore() {
  if (this._dprofIgnore) return;

  this._dprofState.before();
  state = this._dprofState;
}

function asyncAfter() {
  if (this._dprofIgnore) return;
  
  this._dprofState.after();
  state = root;
}

// The root job is done when process.nextTick is called
asyncWrap.disable();
process.nextTick(function () {
  root.after();
});
asyncWrap.enable();

//
// Print result
//

process.on('exit', function () {
  // even though zlib is sync, it still fires async_wrap events,
  // so disable asyncWrap just to be sure.
  asyncWrap.disable();

  const data = {
    'total': timestamp(),
    'version': version,
    'root': root
  };

  if (process.env.NODE_DPROF_DEBUG) {
    fs.writeFileSync('./dprof.json', JSON.stringify(data, null, 1));
  } else {
    fs.writeFileSync('./dprof.json.gz', zlib.gzipSync(JSON.stringify(data)));
  }
});
