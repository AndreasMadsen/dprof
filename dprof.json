{
 "name": "root",
 "init": 0,
 "before": 0,
 "after": 4319266,
 "children": [
  {
   "name": "Timer",
   "init": 2044586,
   "before": 548667,
   "after": 1409916,
   "children": [
    {
     "name": "FSReqWrap",
     "init": 3039113,
     "before": 458225,
     "after": 788615,
     "children": [
      {
       "name": "FSReqWrap",
       "init": 3776193,
       "before": 164991,
       "after": 213630,
       "children": []
      },
      {
       "name": "FSReqWrap",
       "init": 3803875,
       "before": 188243,
       "after": 268189,
       "children": [
        {
         "name": "FSReqWrap",
         "init": 4063866,
         "before": 39850,
         "after": 74510,
         "children": []
        }
       ]
      }
     ]
    },
    {
     "name": "FSReqWrap",
     "init": 3279865,
     "before": 623396,
     "after": 650968,
     "children": []
    }
   ]
  }
 ]
}