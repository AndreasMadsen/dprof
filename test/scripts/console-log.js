'use strict';
require('../../dprof.js');

console.log('hallo world');
