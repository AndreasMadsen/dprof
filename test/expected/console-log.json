{
 "stdout": "hallo world\n",
 "stderr": "",
 "dump": {
  "total": 20461586,
  "version": "0.14.0",
  "root": {
   "name": "root",
   "init": 0,
   "destroy": 19381714,
   "before": [
    0
   ],
   "after": [
    19353590
   ],
   "stack": [
    {
     "description": "Module._compile (module.js:556:32)",
     "filename": "module.js",
     "line": 556,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:565:10)",
     "filename": "module.js",
     "line": 565,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:473:32)",
     "filename": "module.js",
     "line": 473,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:432:12)",
     "filename": "module.js",
     "line": 432,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:424:3)",
     "filename": "module.js",
     "line": 424,
     "collum": 3
    },
    {
     "description": "Module.require (module.js:483:17)",
     "filename": "module.js",
     "line": 483,
     "collum": 17
    },
    {
     "description": "require (internal/module.js:20:19)",
     "filename": "internal/module.js",
     "line": 20,
     "collum": 19
    },
    {
     "description": "Object.<anonymous> (/Users/Jeremiah/Documents/dprof/test/scripts/console-log.js:2:1)",
     "filename": "/test/scripts/console-log.js",
     "line": 2,
     "collum": 1
    },
    {
     "description": "Module._compile (module.js:556:32)",
     "filename": "module.js",
     "line": 556,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:565:10)",
     "filename": "module.js",
     "line": 565,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:473:32)",
     "filename": "module.js",
     "line": 473,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:432:12)",
     "filename": "module.js",
     "line": 432,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:424:3)",
     "filename": "module.js",
     "line": 424,
     "collum": 3
    },
    {
     "description": "Module.runMain (module.js:590:10)",
     "filename": "module.js",
     "line": 590,
     "collum": 10
    },
    {
     "description": "run (bootstrap_node.js:394:7)",
     "filename": "bootstrap_node.js",
     "line": 394,
     "collum": 7
    },
    {
     "description": "startup (bootstrap_node.js:149:9)",
     "filename": "bootstrap_node.js",
     "line": 149,
     "collum": 9
    },
    {
     "description": "bootstrap_node.js:509:3",
     "filename": "bootstrap_node.js",
     "line": 509,
     "collum": 3
    }
   ],
   "children": [
    {
     "name": "Signal",
     "init": 3513179,
     "destroy": null,
     "before": [],
     "after": [],
     "stack": [
      {
       "description": "process.<anonymous> (internal/process.js:195:18)",
       "filename": "internal/process.js",
       "line": 195,
       "collum": 18
      },
      {
       "description": "emitTwo (events.js:106:13)",
       "filename": "events.js",
       "line": 106,
       "collum": 13
      },
      {
       "description": "process.emit (events.js:191:7)",
       "filename": "events.js",
       "line": 191,
       "collum": 7
      },
      {
       "description": "_addListener (events.js:226:14)",
       "filename": "events.js",
       "line": 226,
       "collum": 14
      },
      {
       "description": "process.addListener (events.js:270:10)",
       "filename": "events.js",
       "line": 270,
       "collum": 10
      },
      {
       "description": "Object.<anonymous> (/Users/Jeremiah/Documents/dprof/dprof.js:152:11)",
       "filename": "/dprof.js",
       "line": 152,
       "collum": 11
      },
      {
       "description": "Module._compile (module.js:556:32)",
       "filename": "module.js",
       "line": 556,
       "collum": 32
      },
      {
       "description": "Object.Module._extensions..js (module.js:565:10)",
       "filename": "module.js",
       "line": 565,
       "collum": 10
      },
      {
       "description": "Module.load (module.js:473:32)",
       "filename": "module.js",
       "line": 473,
       "collum": 32
      },
      {
       "description": "tryModuleLoad (module.js:432:12)",
       "filename": "module.js",
       "line": 432,
       "collum": 12
      },
      {
       "description": "Function.Module._load (module.js:424:3)",
       "filename": "module.js",
       "line": 424,
       "collum": 3
      },
      {
       "description": "Module.require (module.js:483:17)",
       "filename": "module.js",
       "line": 483,
       "collum": 17
      },
      {
       "description": "require (internal/module.js:20:19)",
       "filename": "internal/module.js",
       "line": 20,
       "collum": 19
      },
      {
       "description": "Object.<anonymous> (/Users/Jeremiah/Documents/dprof/test/scripts/console-log.js:2:1)",
       "filename": "/test/scripts/console-log.js",
       "line": 2,
       "collum": 1
      },
      {
       "description": "Module._compile (module.js:556:32)",
       "filename": "module.js",
       "line": 556,
       "collum": 32
      },
      {
       "description": "Object.Module._extensions..js (module.js:565:10)",
       "filename": "module.js",
       "line": 565,
       "collum": 10
      },
      {
       "description": "Module.load (module.js:473:32)",
       "filename": "module.js",
       "line": 473,
       "collum": 32
      },
      {
       "description": "tryModuleLoad (module.js:432:12)",
       "filename": "module.js",
       "line": 432,
       "collum": 12
      },
      {
       "description": "Function.Module._load (module.js:424:3)",
       "filename": "module.js",
       "line": 424,
       "collum": 3
      },
      {
       "description": "Module.runMain (module.js:590:10)",
       "filename": "module.js",
       "line": 590,
       "collum": 10
      },
      {
       "description": "run (bootstrap_node.js:394:7)",
       "filename": "bootstrap_node.js",
       "line": 394,
       "collum": 7
      },
      {
       "description": "startup (bootstrap_node.js:149:9)",
       "filename": "bootstrap_node.js",
       "line": 149,
       "collum": 9
      },
      {
       "description": "bootstrap_node.js:509:3",
       "filename": "bootstrap_node.js",
       "line": 509,
       "collum": 3
      }
     ],
     "children": []
    },
    {
     "name": "Pipe",
     "init": 13930532,
     "destroy": null,
     "before": [],
     "after": [],
     "stack": [
      {
       "description": "createHandle (net.js:33:31)",
       "filename": "net.js",
       "line": 33,
       "collum": 31
      },
      {
       "description": "new Socket (net.js:139:20)",
       "filename": "net.js",
       "line": 139,
       "collum": 20
      },
      {
       "description": "createWritableStdioStream (internal/process/stdio.js:156:16)",
       "filename": "internal/process/stdio.js",
       "line": 156,
       "collum": 16
      },
      {
       "description": "process.getStdout [as stdout] (internal/process/stdio.js:10:14)",
       "filename": "internal/process/stdio.js",
       "line": 10,
       "collum": 14
      },
      {
       "description": "console.js:100:37",
       "filename": "console.js",
       "line": 100,
       "collum": 37
      },
      {
       "description": "NativeModule.compile (bootstrap_node.js:497:7)",
       "filename": "bootstrap_node.js",
       "line": 497,
       "collum": 7
      },
      {
       "description": "Function.NativeModule.require (bootstrap_node.js:438:18)",
       "filename": "bootstrap_node.js",
       "line": 438,
       "collum": 18
      },
      {
       "description": "get (bootstrap_node.js:254:34)",
       "filename": "bootstrap_node.js",
       "line": 254,
       "collum": 34
      },
      {
       "description": "Object.<anonymous> (/Users/Jeremiah/Documents/dprof/test/scripts/console-log.js:4:1)",
       "filename": "/test/scripts/console-log.js",
       "line": 4,
       "collum": 1
      },
      {
       "description": "Module._compile (module.js:556:32)",
       "filename": "module.js",
       "line": 556,
       "collum": 32
      },
      {
       "description": "Object.Module._extensions..js (module.js:565:10)",
       "filename": "module.js",
       "line": 565,
       "collum": 10
      },
      {
       "description": "Module.load (module.js:473:32)",
       "filename": "module.js",
       "line": 473,
       "collum": 32
      },
      {
       "description": "tryModuleLoad (module.js:432:12)",
       "filename": "module.js",
       "line": 432,
       "collum": 12
      },
      {
       "description": "Function.Module._load (module.js:424:3)",
       "filename": "module.js",
       "line": 424,
       "collum": 3
      },
      {
       "description": "Module.runMain (module.js:590:10)",
       "filename": "module.js",
       "line": 590,
       "collum": 10
      },
      {
       "description": "run (bootstrap_node.js:394:7)",
       "filename": "bootstrap_node.js",
       "line": 394,
       "collum": 7
      },
      {
       "description": "startup (bootstrap_node.js:149:9)",
       "filename": "bootstrap_node.js",
       "line": 149,
       "collum": 9
      },
      {
       "description": "bootstrap_node.js:509:3",
       "filename": "bootstrap_node.js",
       "line": 509,
       "collum": 3
      }
     ],
     "children": []
    },
    {
     "name": "Pipe",
     "init": 15613167,
     "destroy": null,
     "before": [],
     "after": [],
     "stack": [
      {
       "description": "createHandle (net.js:33:31)",
       "filename": "net.js",
       "line": 33,
       "collum": 31
      },
      {
       "description": "new Socket (net.js:139:20)",
       "filename": "net.js",
       "line": 139,
       "collum": 20
      },
      {
       "description": "createWritableStdioStream (internal/process/stdio.js:156:16)",
       "filename": "internal/process/stdio.js",
       "line": 156,
       "collum": 16
      },
      {
       "description": "process.getStderr [as stderr] (internal/process/stdio.js:23:14)",
       "filename": "internal/process/stdio.js",
       "line": 23,
       "collum": 14
      },
      {
       "description": "console.js:100:53",
       "filename": "console.js",
       "line": 100,
       "collum": 53
      },
      {
       "description": "NativeModule.compile (bootstrap_node.js:497:7)",
       "filename": "bootstrap_node.js",
       "line": 497,
       "collum": 7
      },
      {
       "description": "Function.NativeModule.require (bootstrap_node.js:438:18)",
       "filename": "bootstrap_node.js",
       "line": 438,
       "collum": 18
      },
      {
       "description": "get (bootstrap_node.js:254:34)",
       "filename": "bootstrap_node.js",
       "line": 254,
       "collum": 34
      },
      {
       "description": "Object.<anonymous> (/Users/Jeremiah/Documents/dprof/test/scripts/console-log.js:4:1)",
       "filename": "/test/scripts/console-log.js",
       "line": 4,
       "collum": 1
      },
      {
       "description": "Module._compile (module.js:556:32)",
       "filename": "module.js",
       "line": 556,
       "collum": 32
      },
      {
       "description": "Object.Module._extensions..js (module.js:565:10)",
       "filename": "module.js",
       "line": 565,
       "collum": 10
      },
      {
       "description": "Module.load (module.js:473:32)",
       "filename": "module.js",
       "line": 473,
       "collum": 32
      },
      {
       "description": "tryModuleLoad (module.js:432:12)",
       "filename": "module.js",
       "line": 432,
       "collum": 12
      },
      {
       "description": "Function.Module._load (module.js:424:3)",
       "filename": "module.js",
       "line": 424,
       "collum": 3
      },
      {
       "description": "Module.runMain (module.js:590:10)",
       "filename": "module.js",
       "line": 590,
       "collum": 10
      },
      {
       "description": "run (bootstrap_node.js:394:7)",
       "filename": "bootstrap_node.js",
       "line": 394,
       "collum": 7
      },
      {
       "description": "startup (bootstrap_node.js:149:9)",
       "filename": "bootstrap_node.js",
       "line": 149,
       "collum": 9
      },
      {
       "description": "bootstrap_node.js:509:3",
       "filename": "bootstrap_node.js",
       "line": 509,
       "collum": 3
      }
     ],
     "children": []
    },
    {
     "name": "NextTickWrap",
     "init": 18915862,
     "destroy": 20046371,
     "before": [
      19702040
     ],
     "after": [
      19938171
     ],
     "stack": [
      {
       "description": "onwrite (_stream_writable.js:350:15)",
       "filename": "_stream_writable.js",
       "line": 350,
       "collum": 15
      },
      {
       "description": "WritableState.onwrite (_stream_writable.js:89:5)",
       "filename": "_stream_writable.js",
       "line": 89,
       "collum": 5
      },
      {
       "description": "Socket._writeGeneric (net.js:715:5)",
       "filename": "net.js",
       "line": 715,
       "collum": 5
      },
      {
       "description": "Socket._write (net.js:725:8)",
       "filename": "net.js",
       "line": 725,
       "collum": 8
      },
      {
       "description": "doWrite (_stream_writable.js:307:12)",
       "filename": "_stream_writable.js",
       "line": 307,
       "collum": 12
      },
      {
       "description": "writeOrBuffer (_stream_writable.js:293:5)",
       "filename": "_stream_writable.js",
       "line": 293,
       "collum": 5
      },
      {
       "description": "Socket.Writable.write (_stream_writable.js:220:11)",
       "filename": "_stream_writable.js",
       "line": 220,
       "collum": 11
      },
      {
       "description": "Socket.write (net.js:651:40)",
       "filename": "net.js",
       "line": 651,
       "collum": 40
      },
      {
       "description": "Console.log (console.js:43:16)",
       "filename": "console.js",
       "line": 43,
       "collum": 16
      },
      {
       "description": "Object.<anonymous> (/Users/Jeremiah/Documents/dprof/test/scripts/console-log.js:4:9)",
       "filename": "/test/scripts/console-log.js",
       "line": 4,
       "collum": 9
      },
      {
       "description": "Module._compile (module.js:556:32)",
       "filename": "module.js",
       "line": 556,
       "collum": 32
      },
      {
       "description": "Object.Module._extensions..js (module.js:565:10)",
       "filename": "module.js",
       "line": 565,
       "collum": 10
      },
      {
       "description": "Module.load (module.js:473:32)",
       "filename": "module.js",
       "line": 473,
       "collum": 32
      },
      {
       "description": "tryModuleLoad (module.js:432:12)",
       "filename": "module.js",
       "line": 432,
       "collum": 12
      },
      {
       "description": "Function.Module._load (module.js:424:3)",
       "filename": "module.js",
       "line": 424,
       "collum": 3
      },
      {
       "description": "Module.runMain (module.js:590:10)",
       "filename": "module.js",
       "line": 590,
       "collum": 10
      },
      {
       "description": "run (bootstrap_node.js:394:7)",
       "filename": "bootstrap_node.js",
       "line": 394,
       "collum": 7
      },
      {
       "description": "startup (bootstrap_node.js:149:9)",
       "filename": "bootstrap_node.js",
       "line": 149,
       "collum": 9
      },
      {
       "description": "bootstrap_node.js:509:3",
       "filename": "bootstrap_node.js",
       "line": 509,
       "collum": 3
      }
     ],
     "children": []
    }
   ]
  }
 }
}